services/cd-service/pkg/repository/repository_test.go

```diff
@@ -251,6 +252,81 @@ func TestNew(t *testing.T) {
 				}
 			},
 		},
+		{
+			Name:          "new in empty directory in bootstrap mode works",
+			Setup:         func(_ *testing.T, _, _ string) {},
+			BootstrapMode: true,
+			Test: func(t *testing.T, repo Repository, remoteDir string) {
+				if !repo.State().BootstrapMode {
+					t.Fatal("BootstrapMode was not preserved")
+				}
+			},
+		},
+		{
+			Name: "new in initialized repository in bootstrap mode works",
+			Setup: func(t *testing.T, remoteDir, localDir string) {
+				// run the initialization code once
+				_, err := New(
+					context.Background(),
+					Config{
+						URL:  "file://" + remoteDir,
+						Path: localDir,
+					},
+				)
+				if err != nil {
+					t.Fatal(err)
+				}
+			},
+			BootstrapMode: true,
+			Test: func(t *testing.T, repo Repository, remoteDir string) {
+				state := repo.State()
+				entries, err := state.Filesystem.ReadDir("")
+				if err != nil {
+					t.Fatal(err)
+				}
+				if len(entries) > 0 {
+					t.Errorf("repository is not empty but contains %d entries", len(entries))
+				}
+				if !state.BootstrapMode {
+					t.Fatal("BootstrapMode was not preserved")
+				}
+			},
+		},
+		{
+			Name:          "Reading config in bootstrap mode works",
+			BootstrapMode: true,
+			Setup: func(t *testing.T, remoteDir, localDir string) {
+				EnvironmentConfigsPath := filepath.Join(remoteDir, "..", "environment_configs.json")
+				configContent := `{"uniqueEnv": {"upstream": {"latest": true }}}`
+				err := os.WriteFile(EnvironmentConfigsPath, []byte(configContent), 0644)
+				if err != nil {
+					t.Fatal(err)
+				}
+			},
+			Test: func(t *testing.T, repo Repository, remoteDir string) {
```

Comment:
I think these should be different test-array entirely.
If it uses different "Setup" and "Test" functions, that's an indicator to make it a seperate test-array.

But maybe you can generalize a bit (but since it's so many different things that are happening here, I think making it seperate test-arrays would be simpler)

Reply:
(I know you did invent this pattern. We shouldn't continue using it though)

Reply:
meant to say "did NOT invent"